### PR TITLE
feat: aggregates without group keys

### PR DESCRIPTION
**Pull request by e-dard**
This PR adds a feature that I had already built at a lower level, but forgotten to implement in the Read Buffer's table and external API. Now you can use `read_aggregate` to ask for column aggregates without grouping on any other columns. Predicates are supported. So this effectively provides Read Buffer support for things like:

```sql
SELECT COUNT(temp), SUM(foo), MAX(bar) FROM cpu WHERE time > 100;
```

As part of this work I had to teach tables to merge row group results for `read_aggregate` when there were no group keys. My bad for forgetting this previously.

### CODE REVIEW

File: read_buffer/src/table.rs [hunk 1]
```diff
@@ -1126,7 +1126,77 @@ mod test {
...
+        columns.insert(
+            "region".to_string(),
+            ColumnType::create_tag(&["west", "west", "east"]),
+        );
+        let rg = RowGroup::new(3, columns);
+        let mut table = Table::new("cpu", rg);
+
+        // Build another row group.
+        let mut columns = BTreeMap::new();
+        columns.insert("time".to_string(), ColumnType::create_time(&[2, 3, 4]));
```
Comment (alamb) [inline read_buffer/src/table.rs:1145]: I suggest updating this test so the row groups don't have the same number of rows in them (perhaps this second row group could have 2 rows) to improve boundary condition coverage

**Review by alamb — APPROVED** [review]
I don't understand one of the loop conditions `rows()` seems strange to me -- but otherwise I think this PR makes sense. All in all I vote let's get it in!

File: read_buffer/src/lib.rs [hunk 2]
```diff
@@ -292,6 +292,15 @@ impl Database {
         group_columns: Selection<'input>,
         aggregates: Vec<(ColumnName<'input>, AggregateType)>,
     ) -> Result<ReadAggregateResults> {
+        for (_, agg) in &aggregates {
+            match agg {
+                AggregateType::First | AggregateType::Last => {
+                    return Err(Error::UnsupportedAggregate { agg: *agg });
```
[the remark this answers is not in the document]
Comment (e-dard) [reply read_buffer/src/lib.rs:298]: I just haven't implemented them yet. And I probably won't until I revisit the way I'm doing aggregations more generally in the engine. Perhaps something I will work towards this week.

File: read_buffer/src/row_group.rs [hunk 3]
```diff
@@ -864,6 +872,56 @@ impl RowGroup {
...
+        let aggregate_columns = dst
+            .schema
+            .aggregate_columns
+            .iter()
+            .map(|(col_type, agg_type, _)| (self.column_by_name(col_type.as_str()), *agg_type))
+            .collect::<Vec<_>>();
+
+        let row_ids = match predicate.is_empty() {
+            true => {
+                // TODO(edd): PERF - teach each column encoding how to produce
```
[the remark this answers is not in the document]
Comment (e-dard) [reply read_buffer/src/row_group.rs:887]: Yeah. Basically I will just teach the column to understand `RowIDsOption::All`, which I suppose is very similar!

File: read_buffer/src/row_group.rs [hunk 4]
```diff
@@ -1517,18 +1575,33 @@ impl<'row_group> ReadAggregateResult<'row_group> {
...
     pub fn is_empty(&self) -> bool {
-        self.group_keys.is_empty()
+        self.aggregates.is_empty()
     }
 
     pub fn schema(&self) -> &ResultSchema {
         &self.schema
     }
 
+    // The number of rows in the result.
```
[the remark this answers is not in the document]
Comment (e-dard) [reply read_buffer/src/row_group.rs:1585]: `AggregateResults` does not have to contain group keys. That would be the case if it was a simple aggregate. Therefore the number of aggregate rows is the correct number of rows in the result. When there are group keys present they would have the same number of rows as `aggregates` because there is one aggregate "row" (collection of aggregate values) for each distinct group key.

File: read_buffer/src/row_group.rs [hunk 5]
```diff
@@ -1538,25 +1611,33 @@ impl<'row_group> ReadAggregateResult<'row_group> {
...
-        if !other.group_keys_sorted {
+        if !other.group_keys_sorted() {
             other.sort();
         }
 
-        let self_len = self.cardinality();
-        let other_len = other.cardinality();
+        let self_group_keys = self.cardinality();
+        let other_group_keys = other.cardinality();
+        let self_len = self.rows();
```
[the remark this answers is not in the document]
Comment (e-dard) [reply read_buffer/src/row_group.rs:1633]: `rows` is the number of rows to aggregate. `ReadAggregateResult.aggregates` is a row (vec) storing an aggregate for each aggregate column.

File: read_buffer/src/table.rs (hunk 1 above)
Comment (e-dard) [reply read_buffer/src/table.rs]: Done in 4d107334

File: read_buffer/src/row_group.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (alamb) [reply read_buffer/src/row_group.rs:1585]: @e-dard  I suggest adding this as a comment to the code as it might not be obvious to future readers either

[the remark this answers is not in the document]
Comment (e-dard) [reply read_buffer/src/row_group.rs:1585]: bah I missed this sorry. I've made a note to document it better.

[the remark this answers is not in the document]
Comment (alamb) [reply read_buffer/src/row_group.rs:1585]: I made the comment after you merged the PR -- lol
